tests/parser/base.py

```diff
@@ -0,0 +1,169 @@
+#
+# Blackbox tests of netplan's keyfile parser that verify that the generated
+# YAML files look as expected. These are run during "make check" and
+# don't touch the system configuration at all.
+#
+# Copyright (C) 2021 Canonical, Ltd.
+# Author: Lukas Märdian <slyon@ubuntu.com>
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; version 3.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from configparser import ConfigParser
+import os
+import sys
+import shutil
+import tempfile
+import unittest
+import ctypes
+import ctypes.util
+import contextlib
+import subprocess
+
+exe_generate = os.path.join(os.path.dirname(os.path.dirname(
+    os.path.dirname(os.path.abspath(__file__)))), 'generate')
+
+# make sure we point to libnetplan properly.
+os.environ.update({'LD_LIBRARY_PATH': '.:{}'.format(os.environ.get('LD_LIBRARY_PATH'))})
+
+# make sure we fail on criticals
+os.environ['G_DEBUG'] = 'fatal-criticals'
+
+lib = ctypes.CDLL(ctypes.util.find_library('netplan'))
+
+
+# A contextmanager to catch the output on a low level so that it catches output
+# from a subprocess or C library call, in addition to normal python output
+@contextlib.contextmanager
+def capture_stderr():
+    stderr_fd = 2  # 2 = stderr
+    with tempfile.NamedTemporaryFile(mode='w+b') as tmp:
+        stderr_copy = os.dup(stderr_fd)
+        try:
+            sys.stderr.flush()
+            os.dup2(tmp.fileno(), stderr_fd)
+            yield tmp
+        finally:
+            sys.stderr.flush()
+            os.dup2(stderr_copy, stderr_fd)
+            os.close(stderr_copy)
+
+
+class TestBase(unittest.TestCase):
```

Comment:
Maybe changing the name of this base class to be more representative of what the base is meant for? I mean, currently this is in `test/parser`, so this is more of like testing of various parser functionality. Which is fine. But the `base.py`'s `TestBase` class is typically for testing the keyfile parser. So possibilities: changing `TestBase` to something like `TestKeyfileBase` or changing the `generate()` method to `generate_from_keyfile()` or maybe renaming the `test/parser` directory? Not sure what is best!